### PR TITLE
Stable logging baseline

### PR DESCRIPTION
**Pull request by teddyort**

### CODE REVIEW

File: roboteq_diff_driver/src/driver.cpp [hunk 1]
```diff
@@ -122,6 +129,11 @@ class MainNode
   nav_msgs::Odometry odom_msg;
   ros::Publisher odom_pub;
 
+  roboteq_diff_msgs::WheelRates wheel_rates_msg;
+  roboteq_diff_msgs::WheelRates wheel_rates_desired_msg;
+  ros::Publisher wheel_rate_pub;
+  ros::Publisher wheel_rate_desired_pub;
+
```
Comment (GoldenZephyr) [inline roboteq_diff_driver/src/driver.cpp:136]: standardize wheel rate vs wheel rates

File: roboteq_diff_driver/src/driver.cpp [hunk 2]
```diff
@@ -271,18 +291,31 @@ ROS_DEBUG_STREAM("cmdvel power right: " << right_power << " left: " << left_powe
   else
   {
     // motor speed (rpm)
-    int32_t right_rpm = right_speed / wheel_circumference * 60.0;
-    int32_t left_rpm = left_speed / wheel_circumference * 60.0;
+    // The extra factor of 10.0 is a hack that is necessary for this to work in practice.
+    // Not really clear why we need to do this, but without it the commanded rpm
+    // is too low to make the motor move
+    int32_t right_rpm = right_speed / wheel_circumference * 60.0 * 10.0;
```
Comment (GoldenZephyr) [inline roboteq_diff_driver/src/driver.cpp]: change 10 to a param (?)

File: roboteq_diff_driver/src/driver.cpp [hunk 3]
```diff
@@ -812,7 +905,8 @@ int MainNode::run()
   mstimer = starttime;
   lstimer = starttime;
 
-//  ros::Rate loop_rate(10);
+  // Not sure what the odometry serial rate is, so maybe this should be higher/lower
+  ros::Rate loop_rate(500);
```
Comment (GoldenZephyr) [inline roboteq_diff_driver/src/driver.cpp]: make loop rate a parameter

File: roboteq_diff_driver/src/driver.cpp [hunk 4]
```diff
@@ -536,7 +618,7 @@ void MainNode::odom_loop()
   }
 
   // read sensor data stream from motor controller
-  if (controller.available())
+  while (controller.available() && DELTAT(millis(), nowtime) < 20)
```
Comment (GoldenZephyr) [inline roboteq_diff_driver/src/driver.cpp]: set 20ms to be configurable

**Review by GoldenZephyr — CHANGES_REQUESTED** [review]
[ ] wheel_rate(s)
[ ] scale factor 10 parameter
[ ] loop sleep parameter
[ ] serial timeout parameter

File: roboteq_diff_driver/src/driver.cpp (hunk 1 above)
Comment (GoldenZephyr) [reply roboteq_diff_driver/src/driver.cpp:136]: This change would be a pain and conflicts with some near-future changes that actually remove these wheel rate messages, so it's not worth changing this right now.

File: roboteq_diff_driver/src/driver.cpp [hunk 5]
```diff
@@ -271,18 +303,31 @@ ROS_DEBUG_STREAM("cmdvel power right: " << right_power << " left: " << left_powe
   else
   {
     // motor speed (rpm)
-    int32_t right_rpm = right_speed / wheel_circumference * 60.0;
-    int32_t left_rpm = left_speed / wheel_circumference * 60.0;
+    // The extra factor of 10.0 is a hack that is necessary for this to work in practice.
+    // Not really clear why we need to do this, but without it the commanded rpm
+    // is too low to make the motor move
+    int32_t right_rpm = right_speed / wheel_circumference * 60.0 * roboteq_rpm_scale;
```
Comment (GoldenZephyr) [inline roboteq_diff_driver/src/driver.cpp:309]: Scale factor is now a ros param

File: roboteq_diff_driver/src/driver.cpp [hunk 6]
```diff
@@ -536,7 +630,7 @@ void MainNode::odom_loop()
   }
 
   // read sensor data stream from motor controller
-  if (controller.available())
+  while (controller.available() && DELTAT(millis(), nowtime) < serial_timeout_ms)
```
Comment (GoldenZephyr) [inline roboteq_diff_driver/src/driver.cpp:633]: timeout is now a ros param

File: roboteq_diff_driver/src/driver.cpp [hunk 7]
```diff
@@ -812,7 +917,8 @@ int MainNode::run()
   mstimer = starttime;
   lstimer = starttime;
 
-//  ros::Rate loop_rate(10);
+  // Not sure what the odometry serial rate is, so maybe this should be higher/lower
+  ros::Rate loop_rate(ros_loop_rate);
```
Comment (GoldenZephyr) [inline roboteq_diff_driver/src/driver.cpp:921]: loop rate is now a ros param

**Review by GoldenZephyr — APPROVED** [review]
Fixed the things Teddy and I discussed earlier

**Review by teddyort — COMMENTED** [review]
LGTM Please test if necessary and then merge

**Pull request merged**